File: workspace/src/adapter/http-greeting-api.ts
import type { GreetingApi } from "../contracts/greeting-api";
import type { GreetingResponse } from "../contracts/greeting-response";

export class HttpGreetingApi implements GreetingApi {
  constructor(private readonly baseUrl: string) {}

  async getGreeting(name: string): Promise<GreetingResponse> {
    const url =
      name === ""
        ? `${this.baseUrl}/api/greeting`
        : `${this.baseUrl}/api/greeting?name=${encodeURIComponent(name)}`;

    const response = await fetch(url);

    return (await response.json()) as GreetingResponse;
  }
}

File: workspace/lib/adapter/http_greeting_api.dart
import 'dart:convert';

import 'package:http/http.dart' as http;

import '../contracts/greeting_api.dart';
import '../contracts/greeting_response.dart';

class HttpGreetingApi implements GreetingApi {
  final String baseUrl;
  final http.Client client;

  HttpGreetingApi({required this.baseUrl, http.Client? client})
    : client = client ?? http.Client();

  @override
  Future<GreetingResponse> getGreeting(String name) async {
    final trimmedName = name.trim();
    final uri = trimmedName.isEmpty
        ? Uri.parse('$baseUrl/api/greeting')
        : Uri.parse(
            '$baseUrl/api/greeting?name=${Uri.encodeQueryComponent(trimmedName)}',
          );

    final response = await client.get(uri);
    final json = jsonDecode(response.body) as Map<String, dynamic>;
    return GreetingResponse.fromJson(json);
  }
}
